File: Minimum Fuel Cost to Report to the Capital.cpp
class Solution {
    long long ans=0;
public:
    long long dfs(vector<vector<int>>& roads, int seats,vector<int>*adj,int src,int p)
    {
        long long  count =1;
        for(auto x:adj[src])
        {
            if(x!=p)
            count+=dfs(roads,seats,adj,x,src);
        }
        if(count%seats!=0)
        ans+=(count/seats)+1;
        else ans+=count/seats;
        return count;
    }
    long long minimumFuelCost(vector<vector<int>>& roads, int seats) {
        vector<int>adj[roads.size()+1];
        int n=roads.size();
     using ll=long long;
        for(auto &x:roads)
        {
            adj[x[0]].push_back(x[1]);
            adj[x[1]].push_back(x[0]);
        }
        ll res=0;
        for(auto x:adj[0])
        {
           long long a= dfs(roads,seats,adj,x,0);
           res+=ans;
           ans=0;
        }
    return res;
    }
};

File: Minimum Fuel Cost to Report to the Capital.java
public class Solution {
    
    long fuel = 0;
    
    public long minimumFuelCost(int[][] roads, int seats) {
        List<List<Integer>> graph = new ArrayList<>();
        fuel = 0;
        for (int i = 0; i <= roads.length; i++) graph.add(new ArrayList<>());
        for (int i = 0; i < roads.length; i++) {
            graph.get(roads[i][0]).add(roads[i][1]);
            graph.get(roads[i][1]).add(roads[i][0]);
        }
        
        dfs(graph, seats, 0, 0);
        
        return fuel;    
    }
    
    private int dfs(List<List<Integer>> graph, int seats, int prev, int cur) {
        int people = 1;
        int incoming = 0;
        for (int city: graph.get(cur)) {
            if (city != prev) {
                incoming = dfs(graph, seats, cur, city);
                fuel += (incoming % seats == 0) ? incoming / seats : ( incoming / seats ) + 1;
                people += incoming;
            }
        }
        
        return people;
    }
}
